proxy/src/proxy.rs

```diff
@@ -74,26 +101,33 @@ pub async fn task_main(
 
     let cancel_map = Arc::new(CancelMap::default());
     loop {
-        let (socket, peer_addr) = listener.accept().await?;
-        info!("accepted postgres client connection from {peer_addr}");
-
-        let session_id = uuid::Uuid::new_v4();
-        let cancel_map = Arc::clone(&cancel_map);
-        tokio::spawn(
-            async move {
-                info!("spawned a task for {peer_addr}");
-
-                socket
-                    .set_nodelay(true)
-                    .context("failed to set socket option")?;
-
-                handle_client(config, &cancel_map, session_id, socket).await
+        tokio::select! {
+            accept_result = listener.accept() => {
+                let (socket, peer_addr) = accept_result?;
+                info!("accepted postgres client connection from {peer_addr}");
+
+                let session_id = uuid::Uuid::new_v4();
+                let cancel_map = Arc::clone(&cancel_map);
+                tokio::spawn(
+                    async move {
+                        info!("spawned a task for {peer_addr}");
+
+                        socket
+                            .set_nodelay(true)
+                            .context("failed to set socket option")?;
+
+                        handle_client(config, &cancel_map, session_id, socket).await
+                    }
+                    .unwrap_or_else(|e| {
+                        // Acknowledge that the task has finished with an error.
+                        error!("per-client task finished with an error: {e:#}");
+                    }),
+                );
+            }
+            _ = PROXY_CANCELLATION_TOKEN.cancelled() => {
+                bail!("Exiting");
             }
-            .unwrap_or_else(|e| {
-                // Acknowledge that the task has finished with an error.
-                error!("per-client task finished with an error: {e:#}");
-            }),
-        );
+        }
     }
```

Comment:
Proposing an alternative I've been thinking for pageserver.

It seems this acceptor task would be a good place to have a `tokio::task::JoinSet`, use that to spawn connection handlers. On cancellation, only stop the accepting, but then in a second loop just wait for all tasks to complete. Then when the main task has exited all connections have died.

No atomic counters are needed, but this `tokio::select` will need a separate listening on the joinset to cleanup any already handled connections. Internally it seemed like JoinSet uses a intrusive linked list, so it's implementation should be close to optimal aside from few bugs.

Didn't check that this task is ever awaited for. If it is already, then all should be good.

Reply:
Pseudo-illustrative semi-suggestion:

```rust
pub async fn task_main(...) -> anyhow::Result<()> {

  let connections = tokio::task::JoinSet::new();

  loop {
    tokio::select! {
      accept_result = listener.accept() => {
        ...
        
        connections.spawn( // only changed
          async move {
            ...
          }
        );
      }
      _ = PROXY_CANCELLATION_TOKEN.cancelled() => {
        // stop accepting new connections and start waiting for old to die
        break;
      }
      maybe_exited = connections.join_next() => {
        // maintain the 
        match maybe_exited {
          Some(Ok(())) => { /* normal exit */ },
          Some(Err(e)) if e.is_cancelled() => unreachable!("we do not deschedule tasks which is different from them reacting to cancellation token and deciding to exit themselves"),
          Some(Err(e)) if e.is_panic() => { /* panic has already been logged by panic hook */ },
          Some(Err(e)) => { error!("unexpected error from joined connection: {e:?}); },
          None => { /* this happens when polled an empty connections, as in proxy is idle right now */ }
        }
      }
    }
  }
  
  // we've been asked to stop, let's just wait until connections complete
  while let Some(_) = connections.join_next().await {
    // pretty much the same match as above but I wrote it as an example,
    // can't see what should be done here (assuming no descheduling).
  }
  
  // all done
  Ok(())
}
```

Reply:
This all could be juggled in a single tokio::select! with the preconditions, but I think it's simpler to have more verbose two loops.

Reply:
Probably needs `drop(listener)` as well in the cancellation arm.

Reply:
This has now been implemented, resolving..